### PR TITLE
[service.listenbrainz] 0.2.1

### PR DESCRIPTION
**Pull request by Freso**
### Description

Allows submitting listens to [ListenBrainz](https://listenbrainz.org/).

This is the final version for the Leia branch. Working on a [Python 3-only version](https://gitlab.com/Freso/service.listenbrainz/-/tree/matrix) for submitting for Matrix.

### Checklist:

- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-scripts/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [script.foo.bar] 1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practise but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-scripts/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

File: service.listenbrainz/addon.xml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+		<summary lang="en_US">ListenBrainz listener</summary>
+		<summary lang="sv_SE">ListenBrainz lyssnar</summary>
+		<description lang="da_DK">Lytter for ListenBrainz der sender information om den musik du lytter til i Kodi til ListenBrainz</description>
+		<description lang="en_GB">The ListenBrainz listener will submit info of the music you’re listening to in Kodi to ListenBrainz</description>
+		<description lang="en_NZ">The ListenBrainz listener will submit info of the music you’re listening to in Kodi to ListenBrainz</description>
+		<description lang="en_US">The ListenBrainz listener will submit info of the music you’re listening to in Kodi to ListenBrainz</description>
+		<description lang="sv_SE">Lyssnar för ListenBrainz som skickar information om den musik du lyssnar på i Kodi till ListenBrainz</description>
+		<language></language>
+		<platform>all</platform>
+		<license>GNU GENERAL PUBLIC LICENSE Version 2</license>
```
Comment (basrieter) [inline service.listenbrainz/addon.xml:23]: Please use the license SPDX (See https://spdx.org/licenses/)

File: service.listenbrainz/listener.py [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+        log('nowplaying', SESSION)
+        result = listenbrainz.submit_playingnow(listen)
+        if not result:
+            return
+        # parse response
+        if 'status' in result and result['status'] == 'ok':
+            return
+        elif 'error' in result:
+            msg = result['error']
+            xbmc.executebuiltin(
```
Comment (basrieter) [inline service.listenbrainz/listener.py:113]: Better to use `xbmcgui.Dialog().notification()`

File: service.listenbrainz/listener.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+            write_file(self.file, self.queue)
+            return
+        # parse response
+        if 'status' in result and result['status'] == 'ok':
+            # remove submitted items from the list
+            self.queue = self.queue[50:]
+        elif 'error' in result:
+            code = result['code']
+            msg = result['error']
+            xbmc.executebuiltin(
```
Comment (basrieter) [inline service.listenbrainz/listener.py:172]: Better to use `xbmcgui.Dialog().notification()`

File: service.listenbrainz/utils.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    if puser or ptoken:
+        if puser != user:
+            # username changed
+            pass  # TODO: Validate username
+        if ptoken != token:
+            # token changed
+            listenbrainz.validate_token()
+    elif not (user and token):
+        # no username or token
+        xbmc.executebuiltin(
```
Comment (basrieter) [inline service.listenbrainz/utils.py:54]: Better to use `xbmcgui.Dialog().notification()`

File: service.listenbrainz/utils.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+                requests.exceptions.ConnectionError,
+                ValueError):
+            pass
+        return False
+
+    def validate_token(self):
+        """Validate a ListenBrainz token."""
+        endpoint = 'validate-token'
+        if not self.token:
+            xbmc.executebuiltin(
```
Comment (basrieter) [inline service.listenbrainz/utils.py:171]: Better to use `xbmcgui.Dialog().notification()`

File: service.listenbrainz/utils.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+                'Notification(%s,%s,%i)' %
+                (LANGUAGE(32011), LANGUAGE(32027), 7000))
+            return False
+        payload = {'token': self.token}
+        response = self._get(endpoint, payload=payload).json()
+        if 'message' in response:
+            token_valid = response['message'] == 'Token valid.'
+            if token_valid is False:
+                xbmc.executebuiltin(
```
Comment (basrieter) [inline service.listenbrainz/utils.py:180]: Better to use `xbmcgui.Dialog().notification()`

File: service.listenbrainz/listener.py [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+            videos=videos,
+            radio=radio)
+        # init the monitor class (re-init when settings have changed)
+        self.monitor = MyMonitor(
+            action=self._get_settings,
+            user=user,
+            token=token,
+            server=server)
+        if not listenbrainz.validate_server():
+            xbmc.executebuiltin(
```
Comment (basrieter) [inline service.listenbrainz/listener.py:75]: Better to use `xbmcgui.Dialog().notification()`

File: service.listenbrainz/utils.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+from helpers import is_local  # noqa: E402
+from mapping import kodi_mapping  # noqa: E402
+
+__addon__ = xbmcaddon.Addon()
+__addonid__ = __addon__.getAddonInfo('id')
+__version__ = __addon__.getAddonInfo('version')
+
+HEADERS = {'User-Agent': 'Kodi Media center', 'Accept-Charset': 'utf-8'}
+LANGUAGE = __addon__.getLocalizedString
+DATAPATH = xbmc.translatePath(
```
Comment (basrieter) [inline service.listenbrainz/utils.py:24]: use `xbmcvfs.translatePath`, `xbmc.translatePath` is deprecated.

**Review by basrieter — CHANGES_REQUESTED** [review]
Please review my changes.

**Comment by Freso** [comment]
Unless any of these changes are hard blockers, the proposed changes won’t be making it into the `leia` branch for this addon. I’ll update the code for the `matrix` version.

**Comment by basrieter** [comment]
> Unless any of these changes are hard blockers, the proposed changes won’t be making it into the `leia` branch for this addon. I’ll update the code for the `matrix` version.

Would it not make sense to keep them in sync?

**Comment by Freso** [comment]
It would, and if anyone volunteers to backport changes from `matrix` to `leia`, they’re welcome to do so. I have no interest in dealing with Python 2 any more than I absolutely have to. :)

**Comment by basrieter** [comment]
> It would, and if anyone volunteers to backport changes from `matrix` to `leia`, they’re welcome to do so. I have no interest in dealing with Python 2 any more than I absolutely have to. :)

Ok, I understand. So out of curiosity: why the Leia version then?

**Comment by basrieter** [comment]
BTW: Please do continue on the changes in the Matrix version, so I can merge it.

**Comment by Freso** [comment]
> So out of curiosity: why the Leia version then?

Because development on this add-on started on Leia, so the code is already there.

**Comment by basrieter** [comment]
We had an internal discussion on accepting new Python 2 add-ons in Leia. And in the light of us trying to get rid of Python 2 (which is end of life) and wanting user to move Kodi Matrix which runs Python 3, I will have to reject this PR. With Nexus on its way the era of Python 2 add-ons is almost over. 

I hope you understand our decision and will keeping supporting the Matrix version of your add-on.

**Pull request closed without merging**